### PR TITLE
Courant output capability in MC module

### PR DESCRIPTION
**Pull request by awlostowski-noaa**
This pull request contains additions to Muskingum-Cunge routing model to allow for on the fly calculation and output of the Courant number. The Courant number is an important diagnostic metric of the numerical stability and accuracy of the Muskingum-Cunge routing model (or any numerical model).

### CODE REVIEW

File: notebooks/courant_test.ipynb [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Reach simulation kernel\n",
+    "- Simulate wave propagation along a reach by stepping through time and space, calling the M-C module at each time/space step.\n",
+    "- Store model results in [distance x time] DataFrames\n",
+    "- A Gusssian-shaped lateral inflow forcing is handed to the first segment of the reach. "
```
Comment (jameshalgren) [inline notebooks/courant_test.ipynb:111]: sp. Gaussian

File: notebooks/courant_test.ipynb [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+    "    return flow, depth, velocity, celerity, courant, x_param, ql\n",
+    "        "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## A sensitivity analysis of wave timing & magnitude to C\n",
+    "- call the simulation function above for different timestep sizes. \n",
```
Comment (jameshalgren) [inline notebooks/courant_test.ipynb:229]: uppercase 'C' Call

File: notebooks/courant_test.ipynb [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    "duration = 15 * (60*60)\n",
+    "dx = 2000\n",
+    "rchlen = 20000\n",
+    "q0 = 2000\n",
+    "h0 = 2.78\n",
+    "t_pk = 6 * (60*60)\n",
+    "scale = 0.5 * (60*60)\n",
+    "pk_scale = 0.25\n",
+    "\n",
+    "print(\"starting dt sensitivity simulations, these will take a while, especiall for small dt\")\n",
```
Comment (jameshalgren) [inline notebooks/courant_test.ipynb:289]: sp. especially
"for small dt`.`" (or `...`, or `:`)

**Review by jameshalgren — CHANGES_REQUESTED** [review]
A couple of super fine editorial comments on the text of the notebook.

**Comment by awlostowski-noaa** [comment]
# Review suggestions
To review this pull request, I suggest the following
1. Review changes made to [`MCsingleSegStime_f2py_NOLOOP.py`](https://github.com/awlostowski-noaa/t-route/blob/courant/src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/MCsingleSegStime_f2py_NOLOOP.f90). Here, you will notice an additional subroutine to compute the Courant number and kinematic celerity, the addition of `X` to the `secant2_h` subroutine arguments (`intent(out)`), the addition of `cn`, `ck`, and `X` to `muskingcungenwm` subroutine arguments (`intent(out)`). These changes result in three new return variables when `muskingcungenwm` is called: kinematic celerity, `ck`, the Courant number, `cn`, and the Muskingum-Cunge X parameter, `X`. 
2. Run [`parity_test.ipynb`](https://github.com/awlostowski-noaa/t-route/blob/courant/src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/parity_test.ipynb) via Google Colab. This notebook runs a simple comparison of `muskingcungenwm` returns between the original and edited versions of the M-C FOTRAN module to ensure that changes did not affect the calculations of flow, velocity, or depth. Moreover, the values of kinematic celerity and courant number are independently evaluated to ensure correctness. Finally, a timing test is included, which shows that the wall clock time added by the new `courant` subroutine is negligible at scale. 
3. Test updates to v01 routing by running [`compute_nhd_routing_SingleSeg.py`](https://github.com/awlostowski-noaa/t-route/blob/courant/src/python_routing_v01/compute_nhd_routing_SingleSeg.py), which has been updated to accept additional diagnostic variables from the `muskingcungenwm` subroutine and print those variables to .csv and .nc output files. **NOTE:** minimal testing has been done on this script since additions were made. I specifically suggest running the full suite of test cases and making sure NetCDF output is correct. 
4. Test that v02 routing has not been impacted by changes to the M-C routing module. Instructions/suggestions for understanding changes made to mitigate impacts to v02 and how to test their efficacy are documented in a [markdown file](https://github.com/awlostowski-noaa/t-route/blob/courant/src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/v02_routing_test_instructions.md).

**Comment by awlostowski-noaa** [comment]
Also, once the review is done we will delete the `/courant_dev` directory, which remains largely for diagnostic purposes.

File: src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/parity_test.ipynb [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+    "                params[\"s0\"],\n",
+    "                0,\n",
+    "                params[\"depthp\"]\n",
+    "            )\n",
+    "    end = time.time()\n",
+    "    dev_duration.append(end - start) # seconds\n",
+    "    \n",
+    "    print(\"Completed timing test\", n)\n",
+    "\n",
+    "print(del_t) "
```
Comment (jameshalgren) [inline src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/parity_test.ipynb:396]: This line causes an error -- ?`print(dev_duration)`?

File: src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/v02_routing_test_instructions.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+cd $FORTRAN_RES_DIR
+gfortran varPrecision.f90 -c -O0 -fPIC
+gfortran -c -O2 -fPIC -o module_levelpool.o module_levelpool.f90
+gfortran -c -O2 -fPIC -o pymodule_levelpool.o pymodule_levelpool.f90
+cp $FORTRAN_RES_DIR/*.o $FAST_REACH_DIR 
+
+# define more directory path variables - system specific!
+numpy_I="/home/awlostowski/tr/lib/python3.8/site-packages/numpy/core/include"
+py_I="/usr/include/python3.8"
+py_lib="/home/awlostowski/tr/lib"
```
Comment (jameshalgren) [inline src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/v02_routing_test_instructions.md:33]: We don't want these kinds of home paths even just in instructions. A generic `/home/user/venv/lib` might be better.

File: src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/parity_test.ipynb [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+    "try:\n",
+    "    import google.colab\n",
+    "\n",
+    "    ENV_IS_CL = True\n",
+    "    root = r\"/content/t-route\"\n",
+    "    subprocess.run([\"git\", \n",
+    "                    \"clone\", \n",
+    "                    \"-b\",\n",
+    "                    \"courant\",\n",
+    "                    \"https://github.com/awlostowski-noaa/t-route.git\"])\n",
```
Comment (jameshalgren) [inline src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/parity_test.ipynb:39]: We'll want this to point to the NOAA-OWP master when we make the final commit/merge.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 7]
```diff
@@ -101,8 +101,16 @@ def _handle_args():
...
+    
     root = pathlib.Path("../..").resolve()
     sys.path.append(r"../python_framework_v02")
+    
+    routing_v02_dir = os.path.join(root, "src", "python_routing_v02","fast_reach")
+    sys.path.append(routing_v02_dir)
+    
+    print(root)
+    print(routing_v02_dir)
+
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:114]: It doesn't look like these changes work. Not sure why. I made an updated suggestion that appears to give the path in my recommended edits PR to your branch here. There are a number of additional changes to make in general to the calling structure of the script that we hold for now.

**Comment by jameshalgren** [comment]
Status update: Tests 1, 2, and 4 are complete with various comments and recommended changes. Test 3 underway.

**Review by jameshalgren — CHANGES_REQUESTED** [review]
Please see comments plus additional changes submitted via PR: https://github.com/awlostowski-noaa/t-route/pull/5

Thank you for the detailed review recommendations and excellent work putting this together. 

Also, I was able to spot check to see that the NetCDF files are accurately reproducing the output of previous tests, with the additional variables now present in the files.

File: src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/v02_routing_test_instructions.md (hunk 5 above)
Comment (awlostowski-noaa) [reply src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/v02_routing_test_instructions.md]: roger that. changed,

File: src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/parity_test.ipynb (hunk 6 above)
Comment (awlostowski-noaa) [reply src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/parity_test.ipynb:41]: noted as TODO in the notebook. Will circle back once we are ready to merge.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py (hunk 7 above)
Comment (awlostowski-noaa) [reply src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py]: looks like the changes you made in the PR on my branch worked out. So long as a valid shared object is created in `/fast_reach` tests work.

**Comment by awlostowski-noaa** [comment]
## Timing test shows 15 - 20% speedup of MC module, relative to WRF
I completed the timing test in the final cell of [`parity_test.ipynb`](https://github.com/awlostowski-noaa/t-route/blob/courant/src/fortran_routing/mc_pylink_v00/MC_singleSeg_singleTS/courant_dev/parity_test.ipynb). This analysis shows an average wall clock increase of ~17% over 10 tests of 100000 consecutive function calls.

**Comment by awlostowski-noaa** [comment]
# NOLOOP.f90 is 10% slower than MUSKINGCUNGE.f90, before Courant changes
Prior to the addition of the Courant capability in MC_singleSegStime_f2py_NOLOOP.f90, the module was ~10% slower than the module in MUSKINGCUNGE.f90 currently used in the WRF Hydro model. 

That said, courant and channel geometry subroutines increase wall clock time by 5 - 10% beyond this baseline.

**Comment by jameshalgren** [comment]
> NOLOOP.f90 is 10% slower than MUSKINCUNGE.f90, before Courant changes

@awlostowski-noaa Can you post the code that you used to do the speed test? We may be missing compiler optimizations for  in-lining functions, which would account for at least some of the decreased performance.

**Review by jameshalgren — APPROVED** [review]

**Comment by jameshalgren** [comment]
ping @zhengtaocui

**Pull request merged**